source/opt/trim_capabilities_pass.cpp

```diff
@@ -0,0 +1,197 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/opt/trim_capabilities_pass.h"
+
+#include <cassert>
+#include <functional>
+#include <optional>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include "source/enum_set.h"
+#include "source/opt/ir_context.h"
+#include "source/util/string_utils.h"
+
+namespace spvtools {
+namespace opt {
+
+// Aliases for commonly used types.
+using ExtensionSet = std::unordered_set<std::string>;
+
+// An opcode handler is a function which, given an instruction, returns either
+// the required capability, or spv::Capability::Max. Each handler should
+// determine a single capability. If the same instruction could require multiple
+// capabilities, multiple handler should be implemented.
+using OpcodeHandler = std::function<std::optional<spv::Capability>(
+    const IRContext* context, const Instruction* instruction)>;
+
+// ============== Begin opcode handler implementations. =======================
+//
+// Adding support for a new capability should only require adding a new handler,
+// and updating the
+// kSupportedCapabilities/kUntouchableCapabilities/kFordiddenCapabilities lists.
+//
+// Handlers names follow the following convention:
+//  Handler_<Opcode>_<Capability>()
+
+static std::optional<spv::Capability> Handler_OpReadClockKHR_ShaderClockKHR(
+    const IRContext*, const Instruction*) {
+  // The simple presence of this opcode is enough to determine the capability
+  // requirement.
+  return spv::Capability::ShaderClockKHR;
+}
+
+// Opcode of interest to determine capabilities requirements.
+static const std::unordered_multimap<spv::Op, OpcodeHandler, UnsignedHash>
+    kOpcodeHandlers{
+        {spv::Op::OpReadClockKHR, Handler_OpReadClockKHR_ShaderClockKHR},
+    };
+
+// ==============  End opcode handler implementations.  =======================
+
+CapabilitySet TrimCapabilitiesPass::DetermineRequiredCapabilities() {
+  CapabilitySet required_capabilities;
+  const IRContext* ctx = context();
+
+  get_module()->ForEachInst(
+      [ctx, &required_capabilities](Instruction* instruction) {
+        auto[begin, end] = kOpcodeHandlers.equal_range(instruction->opcode());
+        std::for_each(begin, end,
+                      [&required_capabilities, ctx, instruction](auto& item) {
+                        auto result = item.second(ctx, instruction);
+                        if (result.has_value()) {
+                          required_capabilities.insert(*result);
+                        }
+                      });
+      });
+
+  return required_capabilities;
+}
+
+// Utility function to, given a list of capabilities, retrieve the associated
+// extensions.
+template <typename Container>
+static ExtensionSet GetExtensionsForCapabilities(
+    const Container& capabilities) {
+  ExtensionSet output;
```

Comment:
There is an enum for all extensions (https://github.com/KhronosGroup/SPIRV-Tools/blob/main/source/extensions.h).

Reply:
Oh, and there is an ExtensionToString! Thanks, will use that instead.